src/gd_avif.c

```diff
@@ -0,0 +1,667 @@
+  }
+
+  image = gdImageCreateFromAvifCtx(ctx);
+  // ctx->gd_free(ctx);
+
+  return image;
+}
+
+/*
+  Function: gdImageCreateFromAvifPtr
+
+    See <gdImageCreateFromAvif>.
+
+  Parameters:
+
+    size            - size of Avif data in bytes.
+    data            - pointer to Avif data.
+*/
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvifPtr (int size, void *data)
+{
+  gdImagePtr image;
+  gdIOCtx *ctx = gdNewDynamicCtxEx(size, data, 0);
+
+  if (!ctx)
+    return 0;
+
+  image = gdImageCreateFromAvifCtx(ctx);
+  ctx->gd_free(ctx);
+
+  return image;
+}
+
+/*
+  Function: gdImageCreateFromAvifCtx
+
+    See <gdImageCreateFromAvif>.
+    
+    Additional details: the AVIF library comes with functions to create an IO object from
+    a file and from a memory pointer. Of course, it doesn't have a way to create an IO object
+    from a gdIOCtx. So, here, we use our own helper function, createAvifIOfromCtx().
+
+    Otherwise, we create the image by calling AVIF library functions in order:
+    * avifDecoderCreate(), to create the decoder
+    * avifDecoderSetIO(), to tell libavif how to read from our data structure
+    * avifDecoderParse(), to parse the image
+    * avifDecoderNextImage(), to read the first image from the decoder
+    * avifRGBImageSetDefaults(), to create the avifRGBImage
+    * avifRGBImageAllocatePixels(), to allocate memory for the pixels
+    * avifImageYUVToRGB(), to convert YUV to RGB
+    
+    Finally, we create a new gd image and copy over the pixel data.
+  
+  Parameters:
+
+    ctx              - a gdIOCtx struct
+*/
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvifCtx (gdIOCtx * ctx)
+{
+  int x, y;
+  uint8_t *p8; 
+  uint16_t *p16;
+  gdImage *im = NULL;
+  avifResult result;
+  avifIO *io;
+  avifDecoder *decoder;
+  avifRGBImage rgb;
+
+  decoder = avifDecoderCreate();
+  io = createAvifIOFromCtx(ctx);
+
+  // memset(rgb, 0, sizeof(rgb));
+
+  avifDecoderSetIO(decoder, io);
+
+  result = avifDecoderParse(decoder);
+  if (isAvifError(result, "Could not parse image")) {
+    goto cleanup;
+  }
+
+  // Note again that, for an image sequence, we read only the first image, ignoring the rest.
+  result = avifDecoderNextImage(decoder);
+  if (isAvifError(result, "Could not decode image")) {
+    goto cleanup;
+  }
+
+  // Set up the avifRGBImage. Use default settings unless there's demand to override these.
+  avifRGBImageSetDefaults(&rgb, decoder->image);
+  avifRGBImageAllocatePixels(&rgb);
+
+  result = avifImageYUVToRGB(decoder->image, &rgb);
+  if (isAvifError(result, "gd-avif error: Conversion from YUV to RGB failed")) {
+    goto cleanup;
+  }
+
+  im = gdImageCreateTrueColor(decoder->image->width, decoder->image->height);
+	if (!im) {
+		gd_error("avif error: Could not create GD truecolor image");
+		goto cleanup;
+	}
+
+
+// Read the pixels from the AVIF image and copy them into the GD image.
+// Image depth can be 8, 10, 12, or 16. But if depth>8, pixels are 16-bit.
+
+  if (rgb.depth == 8) {
```

Comment:
Could you please use a switch statement here:

```
swtich (rgb.depth) {
    case 12:
    case 10:
        /* 16 bit conversion */
    case 8:
       /* 8bit conversion */
    default:
        /* error our */
}
```

Reply:
I felt like the comment covered the reasoning here. But you'd prefer if the reasoning was embedded in the code? Or are you more concerned about cases where the image depth would be something other than 8, 10, 12, or 16?

Also, doesn't libavif allow for 16-bit pixel depth?

`uint32_t depth;       // legal depths [8, 10, 12, 16]. if depth>8, pixels must be uint16_t internally`

Reply:
There is no 16bit as it would just be a waste of space.

Reply:
True, I've never seen anything saying AVIF will support 16-bit pixels - just 8, 10, and 12. And yet libavif explicitly says that 16 bits are legal... and, as far as I can tell, 10-bit and 12-bit pixels get stored in 16-bit sizes anyway.

I guess, my point is, if libavif allows for 16-bit pixels, is there harm in gd supporting that?

If you think so, I'll drop it.

Reply:
seems like including 16 costs nothing to gd, so if libavif supports it, not a big deal to leave it in

Reply:
AVIF doesn't support 16bit. We are decoding an AVIF image here and if it says it is 16bit there is probably something wrong. If the bit depth would be set to 0xdeadbeef you would still decode it ...

Reply:
Here we know the avifDecoderNextImage() call succeeded, so decoder->image->depth can only be 8, 10, or 12 (the three bit depths supported by AV1). We also know rgb.depth was set by the avifRGBImageSetDefaults() call to be equal to decoder->image->depth. So rgb.depth can only be 8, 10, or 12.

Note: Although avifRGBImageSetDefaults() can be modified to use different defaults, it is extremely unlikely that the default for rgb.depth will change.

Reply:
Ok! I've implemented the `switch` construction and error message that @cryptomilk suggests.